sycl/unittests/event/EventDestruction.cpp

```diff
@@ -0,0 +1,196 @@
+//==------- EventDestruction.cpp --- Check correct event destruction -------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <helpers/CommonRedefinitions.hpp>
+#include <helpers/PiMock.hpp>
+#include <helpers/TestKernel.hpp>
+
+#include <gtest/gtest.h>
+
+#include <iostream>
+
+using namespace sycl;
+
+static int ReleaseCounter = 0;
+static pi_result redefinedEventRelease(pi_event event) {
+  ++ReleaseCounter;
+  return PI_SUCCESS;
+}
+
+pi_result redefinedMemBufferCreate(pi_context, pi_mem_flags, size_t size,
+                                   void *, pi_mem *,
+                                   const pi_mem_properties *) {
+  return PI_SUCCESS;
+}
+
+class EventDestructionTest : public ::testing::Test {
+public:
+  EventDestructionTest() : Plt{default_selector()} {}
+
+protected:
+  void SetUp() override {
+    if (Plt.is_host() || Plt.get_backend() != backend::opencl) {
+      std::cout << "This test is only supported on OpenCL backend\n";
+      std::cout << "Current platform is "
+                << Plt.get_info<sycl::info::platform::name>();
+      return;
+    }
+
+    Mock = std::make_unique<unittest::PiMock>(Plt);
+
+    setupDefaultMockAPIs(*Mock);
+    Mock->redefine<detail::PiApiKind::piEventRelease>(redefinedEventRelease);
+    Mock->redefine<sycl::detail::PiApiKind::piMemBufferCreate>(
+        redefinedMemBufferCreate);
+  }
+
+protected:
+  std::unique_ptr<unittest::PiMock> Mock;
+  platform Plt;
+};
+
+// Test that events are destructed in correct time
+TEST_F(EventDestructionTest, EventDestruction) {
+  if (Plt.is_host() || Plt.get_backend() != backend::opencl) {
+    return;
+  }
+  sycl::context Context{Plt};
+  sycl::queue Queue{Context, sycl::default_selector{}};
+
+  {
+    ReleaseCounter = 0;
+    sycl::event E1{};
+
+    {
+      sycl::event E0 = Queue.submit([&](cl::sycl::handler &cgh) {
+        cgh.single_task<TestKernel>([]() {});
+      });
+      E1 = Queue.submit([&](cl::sycl::handler &cgh) {
+        cgh.depends_on(E0);
+        cgh.single_task<TestKernel>([]() {});
+      });
+      E1.wait();
+    }
+    EXPECT_EQ(ReleaseCounter, 0);
```

Comment:
Can it be clarified in the comments why these specific values of `ReleaseCounter` should be there?